search-with-geocode/src/main/java/com/esri/arcgismaps/sample/searchwithgeocode/MainActivity.kt

```diff
@@ -0,0 +1,375 @@
+
+        // authentication with an API key or named user is
+        // required to access basemaps and other location services
+        ArcGISEnvironment.apiKey = ApiKey.create(BuildConfig.API_KEY)
+        lifecycle.addObserver(mapView)
+
+        mapView.apply {
+            // set the map to be displayed in the MapView
+            map = ArcGISMap(BasemapStyle.ArcGISStreets)
+
+            // set map initial viewpoint
+            map?.initialViewpoint = Viewpoint(40.0, -100.0, 100000000.0)
+
+            // define the graphics overlay and add it to the map view
+            graphicsOverlays.add(graphicsOverlay)
+
+            // set an on touch listener on the map view
+            lifecycleScope.launch {
+                onSingleTapConfirmed.collect { tapEvent ->
+                    // identify the graphic at the tapped coordinate
+                    val tappedGraphic = identifyGraphic(tapEvent.screenCoordinate)
+                    if (tappedGraphic != null) {
+                        // show the address of the identified graphic
+                        showAddressForGraphic(tappedGraphic)
+                    }
+                }
+            }
+        }
+
+        // once the map has loaded successfully, set up address finding UI
+        lifecycleScope.launch {
+            // load the map then set up UI
+            mapView.map?.load()?.onSuccess {
+                // create the pin symbol
+                pinSourceSymbol = createPinSymbol()
+            }?.onFailure {
+                showError(it.message.toString())
+            }
+        }
+
+        // set the switch to update the isSearchInExtent value
+        extentSwitch.setOnCheckedChangeListener { _, isChecked -> isSearchInExtent = isChecked }
+    }
+
+    override fun onCreateOptionsMenu(menu: Menu): Boolean {
+        menuInflater.inflate(R.menu.menu, menu)
+        val search = menu.findItem(R.id.appSearchBar)
+        // set up address search view and listeners
+        setupAddressSearchView(search.actionView as SearchView)
+        return super.onCreateOptionsMenu(menu)
+    }
+
+    /**
+     * Sets up the address SearchView and uses MatrixCursor to
+     * show suggestions to the user as text is entered.
+     */
+    private fun setupAddressSearchView(addressSearchView: SearchView) {
+        addressSearchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
+            override fun onQueryTextSubmit(address: String): Boolean {
+                // geocode the typed address, search within map's viewpoint as keyword was submitted
+                geocodeAddress(address, true)
+                addressSearchView.clearAndHideKeyboard()
+                return true
+            }
+
+            override fun onQueryTextChange(newText: String): Boolean {
+                // if the newText string isn't empty, get suggestions from the locator task
+                if (newText.isNotEmpty()) {
+                    lifecycleScope.launch {
```

Comment:
In theory, would it be possible to have a race condition here? Suppose I type two letters, after the first letter it goes to the server to get a suggestion, and after the second goes to the server to get a suggestion. If the first suggestion takes a long time and the second suggestion is faster, could the suggestions list update with the second suggestion then, with no further input, suddenly update to the first suggestion?

Even if it's possible, it may be rare enough that it's not worth complicating the sample more.

Reply:
In theory, it would be a race condition here. But in practice, the result returned is always the latest text change as in my testing. So, I will leave it as it is. For samples, we try to stick to code that showcases the SDK functionality and general error handling and leave testing the robustness of the app up to developers.